### PR TITLE
Adding certificate login

### LINKED ISSUE #12: Add certificate logging option
Hello,

I've never worked on developing Actions so I don't know if it's possible but... I would like to improve this one by adding options to allow logging through a certificate path (not sure about this one) or encoded entries.

In a business context, I think it would be relevant to automatically login with credentials others than user / password.

What do you think? If it's fine, I can give it a try 🤗

### PR DESCRIPTION
**Pull request by michaelmaillot**
Closes #12 

Setting every input as optional (in order to be able to choose how to log in).

I was hesitating with adding an extra choice input, indicating authentication type setup (`password`, `certificate` and why not `secret`) but maybe the README file is enough to guide users.

Please let me know what you think regarding this PR, I'll adjust it from your feedback.

### CODE REVIEW

**Comment by garrytrinder** [comment]
Thank you @michaelmaillot we will review shortly 👍

File: README.md
Comment (waldekmastykarz) [inline README.md]: The currently published action is v2.0.1. If we want to use a different version here, shouldn't we increase the version number in package.json?

Since we're adding an extra option to login without breaking the existing functionality, I'd suggest that we move to a v2.1.0 instead of v3.0.0.

Comment (michaelmaillot) [reply README.md]: Oh I didn't see the version number on the package.json, I was only refering to the release notes in the README file.

The fact that all options are now optional (as they were required before) made me think about increasing major version. But I'm fine with increasing minor version.

**Comment by fladurner** [comment]
Would be great if this pull request could be reviewed and merged shortly, we also have the need of doing login using an app id.

File: README.md
Comment (waldekmastykarz) [inline README.md]: If we rename `ADMIN_PASSWORD` to `PASSWORD`, wouldn't that be a breaking change?

File: README.md
Comment (waldekmastykarz) [inline README.md:90]: Let's add an empty line between the heading an the paragraph

File: README.md
Comment (waldekmastykarz) [inline README.md]: _Depending on your tenant environment_

Can we clarify this?

File: README.md
Comment (waldekmastykarz) [inline README.md]: Is there any reason why we suggest naming first 2 secrets starting with lowercase and these 3 secrets with uppercase?

File: dist/index.js
Comment (waldekmastykarz) [inline dist/index.js]: As we're adding tenantId and appId, why not expose them to all login modes rather than just for the certificate?

**Review by waldekmastykarz — CHANGES_REQUESTED** [review]

File: README.md
Comment (michaelmaillot) [reply README.md]: Sure yes. Would that be a problem?

File: README.md
Comment (michaelmaillot) [reply README.md]: No reason at all, just inconstitency 🤷‍♂️

Let me adjust this.

I didn't find any stated naming convention except the Actions REST API [code sample](https://docs.github.com/en/rest/reference/actions#list-organization-secrets--code-samples) which returns full uppercase secrets. So maybe we could follow this?

File: dist/index.js
Comment (michaelmaillot) [reply dist/index.js]: Right, thank you for pointing that out!

File: README.md
Comment (michaelmaillot) [reply README.md]: Maybe I could copy / paste what is said on CLI for Microsoft 365 [login](https://pnp.github.io/cli-microsoft365/cmd/login/) documentation page, except the `CLIMICROSOFT365_TENANT` environment variable part? Like this:

```
ID of the tenant from which accounts should be able to authenticate. Use `common` or `organization` if the app is multitenant.
```

File: README.md
Comment (waldekmastykarz) [reply README.md]: Unless necessary, we should avoid breaking changes as it causes extra work for our users to use the latest version. In this case, I'd suggest that we use two separate variables even if it means a bit more work for us.

File: README.md
Comment (waldekmastykarz) [reply README.md]: That would be better indeed 👍

File: README.md
Comment (waldekmastykarz) [reply README.md]: I'd say, let's align with the existing documentation so that we're not causing any breaking changes. We can always take a note of that to adjust in the next major version.

**Comment by garrytrinder** [comment]
> Would be great if this pull request could be reviewed and merged shortly, we also have the need of doing login using an app id.

Thank you for your comment @fladurner it's great to hear that this new feature will give immediate value for you.

We will be looking to review and, all being well, merge this weekend.

We ask for your patience, but we wanted you to know that this is on our radar 👍🏻

**Comment by fladurner** [comment]
> > Would be great if this pull request could be reviewed and merged shortly, we also have the need of doing login using an app id.
> 
> Thank you for your comment @fladurner it's great to hear that this new feature will give immediate value for you.
> 
> We will be looking to review and, all being well, merge this weekend.
> 
> We ask for your patience, but we wanted you to know that this is on our radar 👍🏻

That sounds great, thank you very much for your work!

**Comment by garrytrinder** [comment]
Thank you for working on the PR @michaelmaillot 

Following a review of this PR over the weekend, I have a concern with the lack of testing that we have on our GitHub Actions and that before we look to merge we should implement unit tests to cover the existing functionality of the action, so that we can have confidence in any future changes.

The lack of unit testing is on us (maintainers), I have raised an issue #15 which I have assigned to myself and I will look into as a priority, once complete we can revisit this PR and make the necessary changes.

**Comment by michaelmaillot** [comment]
Great idea, thank you. Feel free to reach out if you need any help (testing or anything else) 😉

**Comment by garrytrinder** [comment]
I've just merged #16 so we can now look to progress this PR.

@michaelmaillot I've made some changes to the codebase to make it much more testable therefore there will be some updates that need to be made, also note that the `master` branch has been renamed to `main`, to update your local clone execute the below git command.

```sh
git branch -m master main
git fetch origin
git branch -u origin/main main
git remote set-head origin -a
```

**Comment by fladurner** [comment]
Hi everyone, is there any chance that this pull request will be merged within the next days/weeks? Thank you all for your work!

**Comment by michaelmaillot** [comment]
Hi @fladurner I'm working on it, hope to bring something up in the near future 🙂

**Comment by garrytrinder** [comment]
Just checking, are you still working on this @michaelmaillot or is it now ready for review?

**Comment by michaelmaillot** [comment]
Ready for review!

Had to make a couple of tests first, before clicking on the well-known button 😉

**Comment by garrytrinder** [comment]
Awesome 👏🏻 

I'll take a look shortly, let's get this shipped 🚀

File: action.yml
Comment (garrytrinder) [inline action.yml]: Let's align this with the option name and description from `login` command for consistency as it can accept a tenant ID or name.

```suggestion
  TENANT:
    description: "ID of the tenant from which accounts should be able to authenticate. Use common or organization if the app is multitenant. If not specified, use the tenant specified in the 'CLIMICROSOFT365_TENANT' environment variable. If the environment variable is not defined, use 'common' as the tenant identifier"
```

File: action.yml
Comment (garrytrinder) [inline action.yml]: Let's align this with the option name and description from `login` command for consistency.

```suggestion
  APP_ID:
    description: "App ID of the Azure AD application to use for authentication. If not specified, use the app specified in the 'CLIMICROSOFT365_AADAPPID' environment variable. If the environment variable is not defined, use the multitenant PnP Management Shell app"
```

File: action.yml
Comment (garrytrinder) [inline action.yml]: Let's change this back to what is was as UPN is more appropriate than email address for authentication.

```suggestion
    description: "Username (upn) of the admin"
```

File: README.md
Comment (garrytrinder) [inline README.md]: Let's also include the `tenant` option here in the the example as it is required for certificate authentication.

```suggestion
      with:
          TENANT: ${{secrets.TENANT }}
          APP_ID:  ${{ secrets.APP_ID }}
          CERTIFICATE_ENCODED: ${{ secrets.CERTIFICATE_ENCODED }}
          CERTIFICATE_PASSWORD:  ${{ secrets.CERTIFICATE_PASSWORD }}
```

File: README.md
Comment (garrytrinder) [inline README.md]: As `APP_ID` can also be used with password login, let's not suggest that it is just specific to certificate authentication.

Let's create a we have a section that calls out `APP_ID` and `TENANT` separately.

```suggestion
- 2 new secrets if `authType` is `certificate`:
  - `CERTIFICATE_ENCODED` - store the Base64-encoded string of the certificate stored in the Azure AD application
  - `CERTIFICATE_PASSWORD` - store the certificate password

- 2 new secrets if using a custom Azure AD identity:
  - `APP_ID` - store App ID of the Azure AD application to use for authentication
  - `TENANT` - store the ID of the tenant from which accounts should be able to authenticate
 
```

File: README.md
Comment (garrytrinder) [inline README.md]: Let's align this with the option name and description from `login` command for consistency.

```suggestion
- `APP_ID`: App ID of the Azure AD application to use for authentication. If not specified, use the app specified in the `CLIMICROSOFT365_AADAPPID` environment variable. If the environment variable is not defined, use the multitenant PnP Management Shell app
```

File: README.md
Comment (garrytrinder) [inline README.md]: Let's align this with the option name and description from `login` command for consistency.

```suggestion
- `TENANT`: ID of the tenant from which accounts should be able to authenticate. Use `common` or `organization` if the app is multitenant. If not specified, use the tenant specified in the `CLIMICROSOFT365_TENANT` environment variable. If the environment variable is not defined, use `common` as the tenant identifier
```

File: README.md
Comment (garrytrinder) [inline README.md]:
```suggestion
- `authType` is `certificate`: at least `CERTIFICATE_ENCODED` and `APP_ID` are required
```

File: README.md
Comment (garrytrinder) [inline README.md]:
```suggestion
  - `TENANT` will be required. ID of the tenant from which accounts should be able to authenticate. Use `common` or `organization` if the app is multitenant
```

File: README.md
Comment (garrytrinder) [inline README.md]: Lets change this to align with the action input description

```suggestion
- `ADMIN_USERNAME` : Username (upn) of the admin
```

**Review by garrytrinder — CHANGES_REQUESTED** [review]
Great work @michaelmaillot just a few changes to make before we merge this in.

TL;DR;

- Lets update `AAD_APP_ID` and `TENANT_ID` inputs to `APP_ID` and `TENANT` respectively so that the action is consistent with the CLI `login` command
- `APP_ID` and `TENANT` can be used with both `password` and `certificate` login types

**Comment by michaelmaillot** [comment]
Thank you for your review, sounds better when keeping consistency with the CLI options indeed 🙂

**Comment by garrytrinder** [comment]
I've picked this up and will test further later this evening.

**Comment by garrytrinder** [comment]
Merged manually 🚀

**Comment by garrytrinder** [comment]
@fladurner certificate login is now possible in the latest release of the action, https://github.com/marketplace/actions/cli-for-microsoft-365-login thank you for your patience whilst we merged this PR ❤️

**Comment by fladurner** [comment]
> @fladurner certificate login is now possible in the latest release of the action, https://github.com/marketplace/actions/cli-for-microsoft-365-login thank you for your patience whilst we merged this PR ❤️

That's great, thank you so much for your great work!

**Pull request closed without merging**
